app/controllers/users_controller.rb

```diff
@@ -1589,6 +1603,64 @@ def register_second_factor_security_key
     render json: failed_json.merge(error: err.message)
   end
 
+  def create_passkey
+    raise Discourse::NotFound unless SiteSetting.experimental_passkeys
+
+    challenge_session = DiscourseWebauthn.stage_challenge(current_user, secure_session)
+    render json:
+             success_json.merge(
+               challenge: challenge_session.challenge,
+               rp_id: DiscourseWebauthn.rp_id,
+               rp_name: DiscourseWebauthn.rp_name,
+               supported_algorithms: ::DiscourseWebauthn::SUPPORTED_ALGORITHMS,
+               user_secure_id: current_user.create_or_fetch_secure_identifier,
+               existing_passkey_credential_ids: current_user.passkey_credential_ids,
+             )
+  end
+
+  def register_passkey
+    raise Discourse::NotFound unless SiteSetting.experimental_passkeys
+
+    params.require(:name)
+    params.require(:attestation)
+    params.require(:clientData)
```

Comment:
Is this called by the relying server such that it uses camelCase?

Reply:
Correct, `clientData` is embedded in the public key, the browser generates it. 2FA security keys have the same structure (line 1593 in the same controller).